### DISCUSSION TITLE
Path Parameters in Advanced Dependency

### DISCUSSION THREAD

**Discussion by gitnarqt** [fastapi/fastapi#6586] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.get("/{project_id}/team}", response_model=List[schemas.team])
def get_team(
    project_id: UUID,
    is_authorized: bool = Depends(deps.AuthorizationCheck(project_id=project_id))
) -> Any:
# Get the Team through crud operations
 pass

class AuthorizationCheck:
    def __init__(self, project_id) -> None:
        self.project_id = project_id

    def __call__(self, air_data_session: Session = Depends(session),
    auth_id: str = Depends(get_firebase_auth_id)) -> bool:
        if not crud.user.is_user_part_of_module_and_project(
        session,
        project_id=str(self.project_id),
        firebase_auth_id=auth_id,
        ):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="User not part of project",
            )
        return True
```

### Description

I'm implementing a dependency that checks if the user is allowed to access a certain project or not. If they are not allowed, the dependency will raise a 401 exception otherwise, it will return true.

The project_id is provided through the path parameter and several APIs will be using this dependency. Here's an example of one of the API that is supposed to fetch the team working on a specific project api/v2/{project_id}/team

As an attempt, I created a dependency class called AuthorizationCheck that takes as input project_id. However, project_id is not being recognized from the path parameter.

I've read the Advanced Dependency tutorial provided in the documentation but it doesn't mention a way to use the path parameters in it. Is this doable?

### Operating System

Windows

### FastAPI Version

0.54.2

### Python Version

3.9.6

**Comment by frankie567** [comment, marked as answer]
The problem here is that you are using a value that will be available at runtime, `project_id`, at "compile time":

<img width="827" alt="Capture d’écran 2021-09-15 à 07 49 52" src="https://user-images.githubusercontent.com/1144727/133378099-82c58da7-8dec-4405-a8ee-00a219112691.png">

Actually, all you have to do is to change your `__call__` method so that it injects `project_id`. That's the beauty of FastAPI: dependencies are 100% recursive:

```py
@router.get("/{project_id}/team}", response_model=List[schemas.team])
def get_team(
    project_id: UUID,
    is_authorized: bool = Depends(deps.AuthorizationCheck())
) -> Any:
# Get the Team through crud operations
 pass


class AuthorizationCheck:
    def __call__(self, project_id: UUID, air_data_session: Session = Depends(session), auth_id: str = Depends(get_firebase_auth_id)) -> bool:
        if not crud.user.is_user_part_of_module_and_project(
            session,
            project_id=str(project_id),
            firebase_auth_id=auth_id,
        ):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="User not part of project",
            )
        return True
```

If you don't need to manage other parameters, a class may be unnecessary, you can do it with a simple function:

```py
@router.get("/{project_id}/team}", response_model=List[schemas.team])
def get_team(
    project_id: UUID,
    is_authorized: bool = Depends(deps.is_authorized)
) -> Any:
# Get the Team through crud operations
 pass


def is_authorized(self, project_id: UUID, air_data_session: Session = Depends(session), auth_id: str = Depends(get_firebase_auth_id)) -> bool:
    if not crud.user.is_user_part_of_module_and_project(
        session,
        project_id=str(project_id),
        firebase_auth_id=auth_id,
    ):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="User not part of project",
        )
    return True
```

Even better, if you don't need the return value, you can just inject this dependency in the route decorator:

```py
@router.get(
    "/{project_id}/team}",
    response_model=List[schemas.team],
    dependencies=[Depends(deps.is_authorized)],
)
def get_team(project_id: UUID) -> Any:
# Get the Team through crud operations
    pass
```

You can reuse this dependency at will. The only thing you have to take care of is to put the `project_id` parameter in your path pattern. Otherwise, FastAPI will automatically look for it in query parameters.

**Comment by eboddington** [comment]
Hope it's okay to jump into this discussion, is there a recommended way to handle something similar where the argument to the dependency function **is** predefined?

In my case, the authorisation function checks against a specific permission, which will differ between endpoints but is part of the application code, not provided in a request.

Simplified example of what I'd like to achieve:
```python
# oauth2_scheme handles reading the token from the header, this already works well
async def is_authorised(token: str = Depends(oauth2_scheme), permission: str):
    # check the permission here

    if not authorised:
        raise HTTPException(401)  


@router.get("/foo", dependencies=[Depends(is_authorised("read foo"))]):
async def get_foo():
    return "foo"

@router.get("/bar", dependencies=[Depends(is_authorised("read bar"))]):
async def get_bar():
    return "bar"
```

Obviously, the above will not work, but essentially the authorisation logic is the same for every endpoint, it's just the specific permission to check against which changes, and this is known at "compile" time.

**Comment by frankie567** [comment]
@eboddington It's usually better to open your own issue/discussion 😉

For your case, you need a callable class dependency: https://fastapi.tiangolo.com/advanced/advanced-dependencies/#a-callable-instance

The `__init__` method will let you set parameters to tune your dependency, while the `__call__` method will be run in the dependency injection context.

```py
class IsAuthorized:
    def __init__(self, permission: str):
        self.permission = permission

    def __call__(self, token: str = Depends(oauth2_scheme)):
        print(f"Check {self.permission} for token {token}")


@router.get("/foo", dependencies=[Depends(IsAuthorized("read foo"))]):
async def get_foo():
    return "foo"
```

**Comment by eboddington** [comment]
> @eboddington It's usually better to open your own issue/discussion 😉
> 
> For your case, you need a callable class dependency: https://fastapi.tiangolo.com/advanced/advanced-dependencies/#a-callable-instance
> 
> The `__init__` method will let you set parameters to tune your dependency, while the `__call__` method will be run in the dependency injection context.
> 
> ```python
> class IsAuthorized:
>     def __init__(self, permission: str):
>         self.permission = permission
> 
>     def __call__(self, token: str = Depends(oauth2_scheme)):
>         print(f"Check {self.permission} for token {token}")
> 
> 
> @router.get("/foo", dependencies=[Depends(IsAuthorized("read foo"))]):
> async def get_foo():
>     return "foo"
> ```

Sorry I will next time. This looks like exactly what I need though, thanks!

**Comment by gitnarqt** [comment]
@frankie567 Merci beaucoup!! It's exactly what I wanted 🎉
